File: app/src/main/java/com/canli/oya/traininventoryroom/data/repositories/TrainRepository.java
package com.canli.oya.traininventoryroom.data.repositories;

import android.arch.lifecycle.LiveData;

import com.canli.oya.traininventoryroom.data.TrainDatabase;
import com.canli.oya.traininventoryroom.data.entities.TrainEntry;
import com.canli.oya.traininventoryroom.utils.AppExecutors;

import java.util.List;

public class TrainRepository {

    private static TrainRepository sInstance;
    private final TrainDatabase mDatabase;
    private final LiveData<List<TrainEntry>> trainList;
    private final AppExecutors mExecutors;

    private TrainRepository(TrainDatabase database, AppExecutors executors) {
        mDatabase = database;
        mExecutors = executors;
        trainList = loadTrains();
    }

    public static TrainRepository getInstance(TrainDatabase database, AppExecutors executors) {
        if (sInstance == null) {
            synchronized (TrainRepository.class) {
                sInstance = new TrainRepository(database, executors);
            }
        }
        return sInstance;
    }

    private LiveData<List<TrainEntry>> loadTrains(){
        return mDatabase.trainDao().getAllTrains();
    }

    public LiveData<List<TrainEntry>> getTrainList() {
        return trainList;
    }

    public void insertTrain(final TrainEntry train){
        mExecutors.diskIO().execute(new Runnable() {
            @Override
            public void run() {
                mDatabase.trainDao().insertTrain(train);
            }
        });
    }

    public void updateTrain(final TrainEntry train){
        mExecutors.diskIO().execute(new Runnable() {
            @Override
            public void run() {
                mDatabase.trainDao().updateTrainInfo(train);
            }
        });
    }

    public void deleteTrain(final TrainEntry train){
        mExecutors.diskIO().execute(new Runnable() {
            @Override
            public void run() {
                mDatabase.trainDao().deleteTrain(train);
            }
        });
    }

    public LiveData<List<TrainEntry>> getTrainsFromThisBrand(String brandName){
        return mDatabase.trainDao().getTrainsFromThisBrand(brandName);
    }

    public LiveData<List<TrainEntry>> getTrainsFromThisCategory(String category){
        return mDatabase.trainDao().getTrainsFromThisCategory(category);
    }

    public List<TrainEntry> searchInTrains(String query){
        return mDatabase.trainDao().searchInTrains(query);
    }

}

File: app/src/main/java/com/canli/oya/traininventoryroom/data/repositories/TrainRepository.kt
package com.canli.oya.traininventoryroom.data.repositories

import com.canli.oya.traininventoryroom.data.TrainEntry
import com.canli.oya.traininventoryroom.data.datasources.BrandDataSource
import com.canli.oya.traininventoryroom.data.datasources.CategoryDataSource
import com.canli.oya.traininventoryroom.data.datasources.TrainDataSource
import io.reactivex.Flowable

class TrainRepository(private val trainDataSource: TrainDataSource,
                                          private val categoryDataSource: CategoryDataSource,
                                          private val brandDataSource: BrandDataSource) {

    fun getAllTrains(): Flowable<List<TrainEntry>> {
        return trainDataSource.getAllTrains()
    }

    fun getChosenTrainLiveData(trainId : Int): Flowable<TrainEntry> {
        return trainDataSource.getChosenTrainLiveData(trainId)
    }

    suspend fun insertTrain(train: TrainEntry) {
        trainDataSource.insertTrain(train)
    }

    suspend fun updateTrain(train: TrainEntry) {
        trainDataSource.updateTrain(train)
    }

    suspend fun deleteTrain(train: TrainEntry) {
        trainDataSource.deleteTrain(train)
    }

    fun getTrainsFromThisBrand(brandName: String): Flowable<List<TrainEntry>> {
        return trainDataSource.getTrainsFromThisBrand(brandName)
    }

    fun getTrainsFromThisCategory(category: String): Flowable<List<TrainEntry>> {
        return trainDataSource.getTrainsFromThisCategory(category)
    }

    suspend fun searchInTrains(query: String): List<TrainEntry> {
        return trainDataSource.searchInTrains(query)
    }

    fun getAllBrands() = brandDataSource.getAllBrands()

    fun getAllCategories() = categoryDataSource.getAllCategories()
}
